crates/chia-sdk-driver/src/merkle_tree.rs

```diff
@@ -0,0 +1,179 @@
+        }
+
+        let (root, proofs) = MerkleTree::build_merkle_tree(leaves);
+        Self { root, proofs }
+    }
+
+    fn build_merkle_tree(leaves: &[Bytes32]) -> (Bytes32, HashMap<Bytes32, (u32, Vec<Bytes32>)>) {
+        let binary_tree = MerkleTree::list_to_binary_tree(leaves).unwrap();
+        MerkleTree::build_merkle_tree_from_binary_tree(&binary_tree)
+    }
+
+    fn sha256(args: &[&[u8]]) -> Bytes32 {
+        let mut hasher = Sha256::new();
+        args.iter().for_each(|arg| hasher.update(arg));
+
+        let result = hasher.finalize();
+        let result: [u8; 32] = result.into();
+        Bytes32::from(result)
+    }
+
+    fn list_to_binary_tree<T: Clone + Debug>(objects: &[T]) -> Result<BinaryTree<T>, &'static str> {
+        let size = objects.len();
+        if size == 0 {
+            return Err("Cannot build a tree out of 0 objects");
+        }
+        if size == 1 {
+            return Ok(BinaryTree::Leaf(objects[0].clone()));
+        }
+        let midpoint = (size + 1) >> 1;
+        let first_half = &objects[..midpoint];
+        let last_half = &objects[midpoint..];
+        let left_tree = MerkleTree::list_to_binary_tree(first_half)?;
+        let right_tree = MerkleTree::list_to_binary_tree(last_half)?;
+        Ok(BinaryTree::Node(Box::new(left_tree), Box::new(right_tree)))
+    }
+
+    fn build_merkle_tree_from_binary_tree(
+        tuples: &BinaryTree<Bytes32>,
+    ) -> (Bytes32, HashMap<Bytes32, (u32, Vec<Bytes32>)>) {
+        match tuples {
+            BinaryTree::Leaf(t) => {
+                let hash = MerkleTree::sha256(&[HASH_LEAF_PREFIX, t]);
+                let mut proof = HashMap::new();
+                proof.insert(*t, (0, vec![]));
+                (hash, proof)
+            }
+            BinaryTree::Node(left, right) => {
+                let (left_root, left_proofs) = MerkleTree::build_merkle_tree_from_binary_tree(left);
+                let (right_root, right_proofs) =
+                    MerkleTree::build_merkle_tree_from_binary_tree(right);
+
+                let new_root = MerkleTree::sha256(&[HASH_TREE_PREFIX, &left_root, &right_root]);
+                let mut new_proofs = HashMap::new();
+
+                for (name, (path, mut proof)) in left_proofs {
+                    proof.push(right_root);
+                    new_proofs.insert(name, (path, proof));
+                }
+
+                for (name, (path, mut proof)) in right_proofs {
+                    let path = path | (1 << proof.len());
+                    proof.push(left_root);
+                    new_proofs.insert(name, (path, proof));
+                }
+
+                (new_root, new_proofs)
+            }
+        }
+    }
+
+    pub fn get_proof(&self, leaf: Bytes32) -> Option<(u32, Vec<Bytes32>)> {
+        self.proofs.get(&leaf).cloned()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hex_literal::hex;
+    use rstest::rstest;
+
+    /*
+    To generate cases, you can do:
```

Comment:
I'd rather not have this kind of comment in the code but we can leave it if you think it's important context still